feat(examples): awesome gno

matijamarjanovic opened the pull request:
implementation of https://github.com/gnolang/gno/issues/3928

This is an on chain representation of the[ Awesome Gno repo](https://github.com/gnoverse/awesome-gno) which serves a purpose of showcasing dApp built in the community. The way this implementation works is:

- there are 2 lists (avl.Trees) : apps and categories
- app can have multiple categories and multiple apps can be part of a category
- admins (regulated by authz package) can add/update/remove both apps and categories
- apps are onchain representation of real gno-related projects from [awesome-gno](https://github.com/gnoverse/awesome-gno) holding the url to the project and the addresses of the authors

<details>
<summary> UI (mock data): </summary>

![image](https://github.com/user-attachments/assets/1736ebec-c31f-4bd4-8723-2d15ef1a86c4)
![image](https://github.com/user-attachments/assets/9ff65f47-f184-4361-aeca-954b2a58f951)
![image](https://github.com/user-attachments/assets/3045edf7-7623-429c-9dee-1094c09d843d)
![image](https://github.com/user-attachments/assets/39c7dc05-6a0c-4efd-b92a-84350eb02c97)


</details>

matijamarjanovic committed:
initial commit

matijamarjanovic committed:
make good ui

matijamarjanovic committed:
add tests

matijamarjanovic committed:
make some funcs non exported. tidy

matijamarjanovic committed:
don't use string slices but a single string in args of funcs so MsgCall can be used

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)
- [ ] The pull request description provides enough details


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: matijamarjanovic/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🔴 At least 1 user(s) of the organization reviewed the pull request (with state "APPROVED")
    │       ├── 🔴 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🟢 This pull request is a draft
    └── 🟢 Then
        └── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)
    └── 🟢 Not (🔴 Pull request author is user: dependabot[bot])

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

matijamarjanovic committed:
fix links

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

jefft0 commented:
Problem here with "not enough arguments in call to page.Picker", same as https://github.com/gnolang/gno/pull/3793#issuecomment-2740926222 :-)

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
add hof registration & picker implmenetation

matijamarjanovic commented:
> Problem here with "not enough arguments in call to page.Picker", same as [#3793 (comment)](https://github.com/gnolang/gno/pull/3793#issuecomment-2740926222) :-)

Addressed in https://github.com/gnolang/gno/pull/3963/commits/567df81c14e0b67ab65ffa77653a4e6268f3b87d. I've also added hof registration (if you remember https://github.com/gnolang/gno/pull/3479 it caused lint to break in init) which suprisingly causes no trouble here..

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+	return nil
+}
+
+func RemoveAdmin(addr std.Address) error {
+	caller := std.PreviousRealm().Address()
+
+	if !isAdmin(caller) {
+		return ufmt.Errorf("not authorized: only admins can remove admins")
+	}
+	if !addr.IsValid() {
+		return ufmt.Errorf("invalid address")
+	}
+	if admins.Size() <= 1 {
+		return ufmt.Errorf("cannot remove the last admin")
+	}
+
+	admins.Remove(addr.String())
+	std.Emit("AdminRemoved", "address", addr.String())
+	return nil
+}
```
Same as above, use a package for privileged actions. I suggest you dig into p/moul/authz; we are slowly moving away from ownable and authorizable

matijamarjanovic replied:
applied in https://github.com/gnolang/gno/pull/3963/commits/048d3efd444f5e2913af972a477b763d43908e9f

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+	}
+
+	exists := false
+	categories.Iterate("", "", func(key string, value interface{}) bool {
+		cat := value.(Category)
+		if cat.Name == name {
+			exists = true
+			return true
+		}
+		return false
+	})
```
Definitely not like this; you should use the fact that AVL tree has a lookup of O(1) - just use the category name as the key. Make sure that you sanitize the key properly (ie no spaces,etc)

leohhhn replied:
Also you can use `any` instead of `interface{}`

matijamarjanovic replied:
yeah, makes sense, forgot about `avl.Has()`..  implemented in https://github.com/gnolang/gno/pull/3963/commits/4f6f4fa3b7a9e83f8456348b24f3b6f9a0b3ea12

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+		DappIDs:     []uint64{},
+	}
+
+	categories.Set(strconv.FormatUint(nextCategoryID, 10), category)
```
`FormatUint` > `strconv.Itoa`

leohhhn replied:
I meant to change `formatuint` to `itoa`, sorry.

matijamarjanovic replied:
Ok ok, I replaced all ides to be of seqid.ID type so there is no need for conversion anymore anyway. Thank you for advice, for changes se https://github.com/gnolang/gno/pull/3963/commits/8fdefd889c7125c5240a7209927ddbfa8cf012fe

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+	dapps     = avl.NewTree()
+	proposals = avl.NewTree()
+
+	nextDappID     uint64 = 1
+	nextProposalID uint64 = 1
```
seqid is more practical when working with AVL trees

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+	return nil
+}
+
+func ListCategories() []Category {
+	result := []Category{}
+
+	categories.Iterate("", "", func(key string, value interface{}) bool {
+		result = append(result, value.(Category))
+		return false
+	})
+
+	return result
+}
```
If you need this to be a getter for someone outside of this realm, check out avl/rotree, which exposes a tree without setter functions

matijamarjanovic replied:
Yes, that's what I was going for (but also for render), applied in https://github.com/gnolang/gno/pull/3963/commits/6fe1d538187d12017d29c989a9f15d8b40f3562c

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+	return result
+}
+
+func ProposeNewDapp(title, description, url string, categoryNamesStr string) (uint64, error) {
```
For proposals, let's try using some DAO code that already exists. @jeronimoalbi maybe we can integrate with gno.me as we talked, but maybe we should just do commondao+datasource. wdyt?

jeronimoalbi replied:
Using gno.me seems to be the right approach with the caveat that before mainnet will require some small changes to your custom proposal definition methods, there are a couple of changes to be done for security that are still pending. But this way would be interesting so you can have a SubDAO within the gno.me DAO tree.

Using commondao is also valid, as a package or directly though the commondao realm. The first would require a bit more effort. If commondao is used we can also integrate with gnome space if it makes sense and list content there if you support the [datasource interface](https://gno.land/p/jeronimoalbi/datasource$source&file=datasource.gno#L64) from `gno.land/p/jeronimoalbi/datasource`.

matijamarjanovic replied:
Hey, so in https://github.com/gnolang/gno/pull/3963/commits/381c72d24798dc239d4a71f460c776cfca3dbab1 I've added datasource support for Proposals, Dapps and Categories (with tests) as we talked today @leohhhn, I'd love some feedback on that as it is pretty long. Looking forward, I should wait until commonDao package/realm is merged and use that here if I understood our conversation today correctly?

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+		return 0, ufmt.Errorf("URL cannot be empty")
+	}
+
+	var categoryNames []string
+	if categoryNamesStr != "" {
+		for _, name := range strings.Split(categoryNamesStr, ",") {
+			trimmedName := strings.TrimSpace(name)
+			if trimmedName != "" {
+				categoryNames = append(categoryNames, trimmedName)
+			}
+		}
+	}
+
+	if len(categoryNames) > 0 {
+		for _, catName := range categoryNames {
+			categoryExists := false
+			categories.Iterate("", "", func(key string, value interface{}) bool {
+				cat := value.(Category)
+				if cat.Name == catName {
+					categoryExists = true
+					return true
+				}
+				return false
+			})
+			if !categoryExists {
+				return 0, ufmt.Errorf("category '%s' does not exist", catName)
+			}
+		}
+	}
```
Too much nesting, can be simplified

matijamarjanovic replied:
done in https://github.com/gnolang/gno/pull/3963/commits/29f82c1805e64145e9da92f2b85f04a884a245a6

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+	caller := std.PreviousRealm().Address()
+
+	if !isAdmin(caller) {
+		return ufmt.Errorf("not authorized: only admins can update dApp categories")
```
this error repeats, it should be declared somewhere. speciifcally for this, either use the p/ errors for control access (ownable/authz), or declare it at top level of this package.

matijamarjanovic replied:
https://github.com/gnolang/gno/pull/3963/commits/4cd963709ad7c67767e8df221a46cef5e1543955

this commit defines errors in types.gno and uses those

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+
+	dapp := value.(Dapp)
+	for _, catName := range strings.Split(newCategories, ",") {
+		_, found := GetCategory(catName)
+		if !found {
+			return ufmt.Errorf("category not found: %s", catName)
+		}
+	}
+
+	for _, oldCat := range dapp.Categories {
+		found := false
+		for _, newCat := range strings.Split(newCategories, ",") {
+			if oldCat == newCat {
+				found = true
+				break
+			}
+		}
+		if !found {
+			err := RemoveDappFromCategory(dappID, oldCat)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	for _, newCat := range strings.Split(newCategories, ",") {
+		found := false
+		for _, oldCat := range dapp.Categories {
+			if newCat == oldCat {
+				found = true
+				break
+			}
+		}
+		if !found {
+			err := addDappToCategory(dappID, newCat)
+			if err != nil {
+				return err
+			}
+		}
+	}
```
Definitely too much nesting. 

If you need to use double for loops, you're probably doing something wrong - these could be avoided with the proper use of maps/avl trees

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+	caller := std.PreviousRealm().Address()
+
+	if !isAdmin(caller) {
+		return ufmt.Errorf("not authorized: only admins can delete dApps")
```
same comment as above for errors

matijamarjanovic committed:
implement seqid

matijamarjanovic committed:
use addrset instead of avl tree for admins

matijamarjanovic committed:
use authz

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+	admins.Set("g1ej0qca5ptsw9kfr64ey8jvfy9eacga6mpj2z0y", true)
+	admins.Set(std.CurrentRealm().Address().String(), true)
+	admins.Set(std.PreviousRealm().Address().String(), true)
+
```
Hey I figured since authz is already managing admin set so I've just implemented that instead in https://github.com/gnolang/gno/pull/3963/commits/048d3efd444f5e2913af972a477b763d43908e9f

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,508 @@
+	admins.Set(addr.String(), true)
+	std.Emit("AdminAdded", "address", addr.String())
+	return nil
+}
```
applied in https://github.com/gnolang/gno/pull/3963/commits/048d3efd444f5e2913af972a477b763d43908e9f

matijamarjanovic committed:
use seq id for categories as well

matijamarjanovic committed:
implement rotree for listing categories

matijamarjanovic committed:
add error definitions to types and use thos

matijamarjanovic committed:
authz doesnt return boolean but err, fix

matijamarjanovic committed:
simplify code in ProposeNewDapp

matijamarjanovic committed:
more simplifying + add new error type

matijamarjanovic committed:
add errors, fix minor bugs

matijamarjanovic committed:
fmt

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
add working tests

matijamarjanovic committed:
add datasource integration

matijamarjanovic committed:
add comment separators

matijamarjanovic committed:
fmt

matijamarjanovic committed:
remove print

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
Merge branch 'feat/awesome-gno' of https://github.com/matijamarjanovic/gno into feat/awesome-gno

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,475 @@
+
+	categories = avl.NewTree()
+
+	auth *authz.Authorizer
+)
+
+func init() {
+	auth = authz.New()
```
```suggestion
	auth = authz.New()
)

func init() { 
```

leohhhn replied:
Initialize it in the var block

matijamarjanovic replied:
done in https://github.com/gnolang/gno/pull/3963/commits/71484f579073c43cae1e6bd46d9b3c315ffd4bdf

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,475 @@
+
+func RemoveAdmin(addr std.Address) error {
+	return auth.Do("remove_admin", func() error {
+		if !addr.IsValid() {
+			return ErrInvalidAddress
+		}
```
Why check for this? Makes sense for adding as you don't want to add an invalid address, but for the removal its the user's fault

matijamarjanovic replied:
makes sense, removed in https://github.com/gnolang/gno/pull/3963/commits/71484f579073c43cae1e6bd46d9b3c315ffd4bdf

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,475 @@
+		if !addr.IsValid() {
+			return ErrInvalidAddress
+		}
+		if auth.Current().(*authz.MemberAuthority).Tree().Size() <= 1 {
+			return ErrCannotRemoveLastAdmin
+		}
+		if err := auth.Current().(*authz.MemberAuthority).RemoveMember(addr); err != nil {
+			return err
+		}
```
It's generally nice to leave a newline after each code block; some people do it some don't. Choose one, and be consistent :)

matijamarjanovic replied:
added in line 61 as well as after other code blocks in the realm where it was missing in https://github.com/gnolang/gno/pull/3963/commits/71484f579073c43cae1e6bd46d9b3c315ffd4bdf

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
- init auth in var block
- add comment describing whose address it is
- remove the realms as admins
- remove address check for removing from admin
- add new lines after code blocks where they are missing

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,475 @@
+func init() {
+	auth = authz.New()
+
+	auth.Current().(*authz.MemberAuthority).AddMember(std.Address("g1ej0qca5ptsw9kfr64ey8jvfy9eacga6mpj2z0y"))
```
makes sense, done in https://github.com/gnolang/gno/pull/3963/commits/71484f579073c43cae1e6bd46d9b3c315ffd4bdf

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,475 @@
+
+	auth.Current().(*authz.MemberAuthority).AddMember(std.Address("g1ej0qca5ptsw9kfr64ey8jvfy9eacga6mpj2z0y"))
+	auth.Current().(*authz.MemberAuthority).AddMember(std.CurrentRealm().Address())
+	auth.Current().(*authz.MemberAuthority).AddMember(std.PreviousRealm().Address())
```
i honestly don't remember the thought process which led to this, so i removed it in https://github.com/gnolang/gno/pull/3963/commits/71484f579073c43cae1e6bd46d9b3c315ffd4bdf

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,475 @@
+	}
+
+	category := Category{
+		ID:          uint64(nextCategoryID.Next()),
```
Yes I've noticed this but I am doing all the conversion since using seqid returns a string that looks like this: "000001" so i figured it might be easier to convert it to numbers this way. But now that I think about it there is no need for id to be a number I can just use the string representation; I will be updating the code shortly

matijamarjanovic replied:
update applied in https://github.com/gnolang/gno/pull/3963/commits/8fdefd889c7125c5240a7209927ddbfa8cf012fe

matijamarjanovic committed:
convert all uint64 ids to seqid; stop using strconv in general

matijamarjanovic committed:
fix fmt

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+		return 0, ErrCategoryExists
+	}
+
+	sanitizedKey := strings.TrimSpace(name)
```
I think we just use seqid, and then make a dashboard page with txlinks that will allow editing of the category and its items (ie preset the ID in the txlink for each category). Check how I did something similar for `r/gnoland/events`

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+		return 0, ErrCategoryExists
+	}
+
+	category := Category{
```
you should set the value in the avl to be a pointer to the memory, its easier to update later.

matijamarjanovic replied:
yeah, done in https://github.com/gnolang/gno/pull/3963/commits/4a7678839c44a530f828c8aca9260549d92fec3a

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+		Name:        name,
+		Description: description,
+		CreatedAt:   time.Now(),
+		DappIDs:     []seqid.ID{},
```
why not just a full avl tree of the dapps themselves? I did something similar in the Hall of Fame

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+
+	std.Emit("CategoryAdded",
+		"id", category.ID.String(),
+		"name", name)
```
```suggestion
"name", name,
)
```

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+		"id", category.ID.String(),
+		"name", name)
+
+	return category.ID, nil
```
Also, let's work towards more and more top-level functions panicking instead of returning errors. this is a general way to go forward: errors at top level functions should only be used with clear intent, the default should be a panic

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+	return nil
+}
+
+func ListCategories() []Category {
```
Why not just expose the rotree? whoever wants to have a slice after can do so on their own

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+	return result
+}
+
+func ProposeNewDapp(title, description, url string, categoryNamesStr string) (seqid.ID, error) {
```
Same thing for errors vs panics.

On top of this, let's think about the way people would use this. let's try to make this app as user-friendly as possible - with txlinks, since we have adena support now. 

`categoryNamesStr` is a workaround; try to think about some option that will make is super easy to propose an app, and maybe then later add it to more categories, like tags for example.

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+		}
+	}
+
+	dapp := Dapp{
```
Same thing for pointers and trees

matijamarjanovic replied:
Yep, implemented in https://github.com/gnolang/gno/pull/3963/commits/4a7678839c44a530f828c8aca9260549d92fec3a

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+	return nil
+}
+
+func GetDapp(id seqid.ID) (Dapp, bool) {
```
ideally we work on making these easier with the rotree. on top of this, I'm not sure we need all of these getters in the first place - i think we need only direct actions on top of Dapps, ie add delete edit, etc.

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+		return 0, ErrNoCategoriesProvided
+	}
+
+	categoryNames := []string{}
```
These for loops should generally be avoided

leohhhn reviewed (changes requested):
This is a good start 👍 

I left some comments before we can continue. I think we should change the direction of the code a bit, try using more Gno best practices, and try to make the app as user-friendly as possible with txlinks + Adena.

On top of this, I know you spent time implementing the proposal feature, but I think that for now we should start simple - scrap this and just allow admins to add/remove/update items - until the [commondao PR](https://github.com/gnolang/gno/pull/3882) is merged. After that, we can make a new PR to implement it here.

We need more dogfooding on the DAO topic, and I wouldn't like to see a milion ways to make a proposal/voting system.

matijamarjanovic commented:
> On top of this, I know you spent time implementing the proposal feature, but I think that for now we should start simple - scrap this and just allow admins to add/remove/update items - until the [commondao PR](https://github.com/gnolang/gno/pull/3882) is merged. After that, we can make a new PR to implement it here.

I was aware this might happen when I was making the system so don't worry about it, I completely agree. I will be addressing the changes and ping you when I'm done. Thank you for the insight

matijamarjanovic committed:
remove add/remove admin and expose Auth object, which by itself is safe to expose and allows admins to manage privileges

matijamarjanovic committed:
Merge branch 'feat/awesome-gno' of https://github.com/matijamarjanovic/gno into feat/awesome-gno

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+	hof.Register("AwesomeGno", "Awesome Gno is a registry for gnowesome dApps on gno.land")
+}
+
+func AddAdmin(addr std.Address) error {
```
> you can also technically expose the Auth object (please double check this), which will expose the transfer functionality into which we can plug in a new authority afterwards.

Correct, this was done in https://github.com/gnolang/gno/pull/3963/commits/e9b63d45391711ee16f7ebff69c118cb43122365

matijamarjanovic committed:
add new errror and use it correctly

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,474 @@
+		return 0, ErrNotAuthorized
+	}
+
+	if name == "" {
```
yeah, updated in https://github.com/gnolang/gno/pull/3963/commits/ef9ba543f694f838d4e1d54b42405ffe795d3e10

matijamarjanovic committed:
instead of saving objects in the tree, save pointers

matijamarjanovic committed:
change types to dapp pointers

matijamarjanovic committed:
- remove all errors and use panics
- get rid of data source, a new one should be made since there are going to be made changes in data types

matijamarjanovic committed:
rewrite the whole code with basic operations and using authz correctly. remove proposals as a concept

matijamarjanovic committed:
add tests

matijamarjanovic committed:
add datasource support

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
add txlink support

matijamarjanovic commented:
The more I got into implementing changes you requested @leohhhn the more I realized that this had to be redone, so the last couple of commits contain:

- removal of the proposal making/aproving concept
- only add/update/delete dapps and categories are left
- dapps and categories structs are containing avl trees of eachother instead of slices of ids
- full capacity of authz: all the exposed funcs are using the `Do` function for authz that calls non-exposed functions with actual logic. this way when common dao is available it can be more easily implemented by just calling these functions and until then admins can manage the realm
- introduction of panics instead of errors
- all avl.Trees are using seqids as keys
- exposed RO trees in place of getters
- minimalistic and better render using txlink for better ux
- datasource covering new data

in short terms, all the requested changes have been applied 1 by 1 and although there were a lot of changes i'd say this is a much cleaner and more readable version

matijamarjanovic committed:
allow app to have multiple authors

leohhhn reviewed (approved):
Ok, I left a couple of final comments. 

Let's merge this, work on adding commondao, and then move it to a more official namespace.

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,268 @@
+	}); err != nil {
+		panic(err)
+	}
+}
+func DeleteCategory(categoryID seqid.ID) {
```
```suggestion
}

func DeleteCategory(categoryID seqid.ID) {
```

check for formatting

matijamarjanovic replied:
fixed in https://github.com/gnolang/gno/pull/3963/commits/f957103aa49a2ed6d8202e7e9352ab6d025d1b76

leohhhn commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,268 @@
+		Authors:     []std.Address{author},
+		URL:         url,
+		CreatedAt:   time.Now(),
+		categories:  avl.NewTree(),
```
And this; not sure you need to make it double linked line this. You can probably just have a slice of strings, like say, tags, that you could list. these would match the category names/IDs so they would be clickable. we do the same thing in p/demo/blog.

Trees make sense when we expect a large amount of items to store. in this case, i don't see this going over 10 or 15 at maximum

leohhhn commented on examples/gno.land/r/demo/awesomegno/render.gno:
```diff
@@ -0,0 +1,250 @@
+		out += md.Italic("No dapps in this category yet")
+	}
+
+	out += md.H2("Management")
+	out += md.Paragraph("Use these actions to manage the category:")
```
rendering will fail here without a newline

matijamarjanovic replied:
![image](https://github.com/user-attachments/assets/d8295957-461f-45bd-9bda-0b2995199dd2)

unless i'm seeing something, seems fine to me. regardless, added the new line just in case in https://github.com/gnolang/gno/pull/3963/commits/7c90ef5dba2b0edf2690f5f90481358f7780dadf

leohhhn commented on examples/gno.land/r/demo/awesomegno/render.gno:
```diff
@@ -0,0 +1,250 @@
+
+func renderHome() string {
+	out := md.H1("AwesomeGno")
+	out += md.Paragraph("Welcome to AwesomeGno - A curated list of awesome dapps built on Gno.land. Here you can discover and explore the growing ecosystem of decentralized applications.")
```
Also, i think it makes sense that the home page loads some apps. maybe the most recent ones, or something similar. 

it also seems you have two "about" sections. This, and the one at the bottom of the page.

matijamarjanovic replied:
casing fixed in https://github.com/gnolang/gno/pull/3963/commits/99dc36ec4b86541e65409e95bf91e459863176f6

matijamarjanovic replied:
on homepage, removed "navigation" section and instead displayed 5 most recently added projects and added "see all" (also deprioritezed quick stats - maybe not necessary at all?):

![image](https://github.com/user-attachments/assets/3e0390e4-89e9-4b6d-8d39-20310f730db6)

done in https://github.com/gnolang/gno/pull/3963/commits/c29a29730c3280dd02617e10cdc89528bf132e19

leohhhn commented on examples/gno.land/r/demo/awesomegno/render.gno:
```diff
@@ -0,0 +1,250 @@
+	out += md.BulletList(links)
+
+	out += md.H2("About")
+	out += md.Paragraph("AwesomeGno is a community-driven project that aims to showcase and promote the best dapps built on Gno.land. Each dapp is carefully categorized and reviewed to ensure quality and usefulness.")
```
maybe this is more of a footnote/disclaimer?

matijamarjanovic replied:
makes more sense, i made it a "blockquote" 
![image](https://github.com/user-attachments/assets/a89b5ca4-2fcc-4132-8178-3add2263c9c4)

in https://github.com/gnolang/gno/pull/3963/commits/fbb2610dc66bbce2ff8fb2b7d0d58a947f6ac660 (i left it at the bottom, idk if this is the best choice, made most sense)

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
- change dapps to items
- some fmt fixes
- chang hof to hor
- make author address optional

matijamarjanovic committed:
some formatting

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,268 @@
+	"gno.land/p/demo/seqid"
+	"gno.land/p/moul/authz"
+	"gno.land/r/leon/hof"
+)
```
okay done in https://github.com/gnolang/gno/pull/3963/commits/f957103aa49a2ed6d8202e7e9352ab6d025d1b76

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,268 @@
+)
+
+func init() {
+	hof.Register("AwesomeGno", "Awesome Gno is a registry for gnowesome dApps on gno.land")
```
done in https://github.com/gnolang/gno/pull/3963/commits/05a891e2794a42ae56a9048ebb7dd940b39abc99

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,268 @@
+	hof.Register("AwesomeGno", "Awesome Gno is a registry for gnowesome dApps on gno.land")
+}
+
+func AddDapp(title, description, url, authorAddress string) seqid.ID {
```
Ok, implemented in https://github.com/gnolang/gno/pull/3963/commits/05a891e2794a42ae56a9048ebb7dd940b39abc99 (i have only removed the address validation but user still needs to pass an empty string at least)

matijamarjanovic committed:
remove uppercase in gno.land

matijamarjanovic committed:
make exported adds not return anything

matijamarjanovic committed:
add newline

matijamarjanovic committed:
make the bottom text more of a disclaimer

matijamarjanovic committed:
display couple of most recent items on homepage

matijamarjanovic committed:
remove my addresses from authorized people and remove mock

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,268 @@
+	nextDappID     seqid.ID
+	nextCategoryID seqid.ID
+
+	authority = authz.NewMemberAuthority(
+		std.Address("g125em6arxsnj49vx35f0n0z34putv5ty3376fg5"), // leon
+		std.Address("g1manfred47kzduec920z88wfr64ylksmdcedlf5"), // moul
+		std.Address("g1ej0qca5ptsw9kfr64ey8jvfy9eacga6mpj2z0y"), // matijamarjanovic (for testing)
+	)
+	Auth = authz.NewWithAuthority(authority)
```
Implementation of this will always result in "unathorized" panic since triggering the `Do` method of Auth checks for `PreviosRealm()` which is in this case always `awesomegno`. This is not the case if `authority` object is initialized in awesomegno realm (like it is already)

matijamarjanovic committed:
make items have slices of category ids instead of using avl.Trees

matijamarjanovic committed:
fix fmt

matijamarjanovic commented on examples/gno.land/r/demo/awesomegno/mock.gno:
removed in 20d5c74a31b276b4b81e9cd176af762c1189df8c

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

leohhhn committed:
Merge branch 'master' into feat/awesome-gno

leohhhn commented:
Let's see if the CI is ok with the code (it won't be) - try to fix up the tests and test it manually to support cross/crossing.

if not, we can merge into examples_draft and then you can make a PR to fix it later

matijamarjanovic committed:
update awesomegno with new authz and crossing

matijamarjanovic committed:
move the `crossing()` to the top of the function being called by another realm from the top of the func to be called in p/authz

matijamarjanovic committed:
fix fmt

matijamarjanovic committed:
move awesome to examples_draft

matijamarjanovic committed:
add error explanation

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

leohhhn committed:
Merge branch 'master' into feat/awesome-gno

moul commented on examples_drafts/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,315 @@
+	cross(hor.Register)("AwesomeGno", "Awesome Gno is a registry for gnowesome dApps, tools and resources on gno.land")
+}
+
+func AddItem(title, description, url, authorAddress string) {
+	crossing()
+
+	if title == "" {
+		panic("item title cannot be empty")
+	}
+
+	addr := std.Address(authorAddress)
+	if err := Auth.DoByPrevious("add_item", func() error {
+		addItem(title, description, url, addr)
```
```suggestion
func AddItem(title, description, url, author std.Address) {
	crossing()

	if title == "" {
		panic("item title cannot be empty")
	}

	if err := Auth.DoByPrevious("add_item", func() error {
		addItem(title, description, url, author)
```

or maybe just `author := std.PreviousRealm().Address()`.

matijamarjanovic replied:
Hey I'd rather keep the primitive type argument so at least this func can be triggered with MsgCall tx (I've changed the variable name to author in https://github.com/gnolang/gno/pull/3963/commits/ce0ccc15b727a6b0c2e300a1b0e07f53c0305272) if that is okay

matijamarjanovic committed:
change variable name from addr to author as per suggested

leohhhn commented on examples_drafts/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,315 @@
+	}
+}
+
+func UpdateItem(itemID seqid.ID, title, description, url string, categories ...seqid.ID) {
```
txlinks for this will not work; MsgCall needs primitive types and variadic is represented as a slice under the hood. 

I'm not sure if we have collapsible sections currently; if we do, i suggest that we have a "add categories" section on each item page or similar, which expands to show all possible categories, where each one is linked with a txlink to add that specific one to the item

matijamarjanovic replied:
I have added UpdateItemCategory function that updates 1 category at a time usable by tx link in https://github.com/gnolang/gno/pull/3963/commits/482ac2c9c1e5cbe1d03e3aae6eb6dd18920b47be. 

This is a good idea and I'm not sure I've seen collapsible sections in render but I'll check and if there is none I can try and add it myself if that is okay with the team

leohhhn reviewed (commented):
Trying the app out a bit manually, I see a few more thinks that could be better; try using the app manually a bit and see where it can be improved

matijamarjanovic committed:
move back to /demo and remove disclaimer

matijamarjanovic commented on examples_drafts/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,315 @@
+/*
+ DISCLAIMER:
+ All exported functions in this realm when called, panic wih the same error message.
+ Output returned by using MsgCall tx:
+
+ gnokey maketx call
+ / -pkgpath "gno.land/r/demo/awesomegno"
+ /-func "AddItem"
+ /-args "title"
+ /-args "desc"
+ /-args "url"
+ /-args "g1kjsl2ungmc95mgluq96w8dqlep8d4n8cxdfthk"
+ /-gas-fee 1000000ugnot
+ /-gas-wanted 500000000
+ /-broadcast
+ /-chainid "dev"
+ /-remote "tcp://127.0.0.1:26657"
+ mainwallet
+
+ Enter password.
+ TX HASH:    
+ INFO:       vm.version=develop
+ --= Error =--
+ Data: should not happen, should be heapItemType: title<()~VPBlock(1,0)>      
+ Msg Traces:
+...
+*/
+package awesomegno
```
you're right, works fine now, moved back to r/demo and removed the disclaimer comment in https://github.com/gnolang/gno/pull/3963/commits/9d40aacaa0528b4630ac637c3b998d106c257faf

matijamarjanovic committed:
Update examples_drafts/r/demo/awesomegno/render.gno

Co-authored-by: Leon Hudak <33522493+leohhhn@users.noreply.github.com>

matijamarjanovic committed:
Update examples_drafts/r/demo/awesomegno/render.gno

Co-authored-by: Leon Hudak <33522493+leohhhn@users.noreply.github.com>

matijamarjanovic committed:
Update examples_drafts/r/demo/awesomegno/render.gno

Co-authored-by: Leon Hudak <33522493+leohhhn@users.noreply.github.com>

matijamarjanovic committed:
Update examples_drafts/r/demo/awesomegno/render.gno

Co-authored-by: Leon Hudak <33522493+leohhhn@users.noreply.github.com>

matijamarjanovic committed:
add author addess validation for add & update item

matijamarjanovic commented on examples_drafts/r/demo/awesomegno/awesomegno.gno:
```diff
@@ -0,0 +1,315 @@
+		ID:          nextItemID.Next(),
+		Title:       name,
+		Description: description,
+		Authors:     []std.Address{author},
```
added address validation in exported funcs AddItem and UpdateItem in https://github.com/gnolang/gno/pull/3963/commits/a53e4981153856d5c8349de875c63d863cd2e213

matijamarjanovic replied:
Makes sense, removed the check and added some new lines and made the authors render registered name if there is one and also made it clickable in https://github.com/gnolang/gno/pull/3963/commits/6af6987c9d7440178654fd2940f07826308e972c

matijamarjanovic committed:
move awesomegno to r/demo, update is valid addr and add update item category function

matijamarjanovic committed:
date formatting, add dot, add new lines and change block to italic as requested

matijamarjanovic commented on examples_drafts/r/demo/awesomegno/render.gno:
```diff
@@ -0,0 +1,285 @@
+				"author", "",
+			).URL()
+		out += md.Paragraph(md.Link("➕ Add New Item", addItemLink))
+		out += "\n"
```
done in https://github.com/gnolang/gno/pull/3963/commits/80e773d5431834f3d59c5999a84302832671f079

matijamarjanovic commented on examples_drafts/r/demo/awesomegno/render.gno:
```diff
@@ -0,0 +1,285 @@
+	if len(itemsList) > 0 {
+		out += md.BulletList(itemsList)
+	} else {
+		out += md.Italic("No items in this category yet")
```
added in https://github.com/gnolang/gno/pull/3963/commits/80e773d5431834f3d59c5999a84302832671f079

matijamarjanovic commented on examples_drafts/r/demo/awesomegno/render.gno:
```diff
@@ -0,0 +1,285 @@
+			}
+			return strings.Join(itemAuthors, ", ")
+		}()),
+		ufmt.Sprintf("Created: %s", item.CreatedAt.Format("2006-01-02")),
```
formatted in https://github.com/gnolang/gno/pull/3963/commits/80e773d5431834f3d59c5999a84302832671f079

matijamarjanovic commented on examples_drafts/r/demo/awesomegno/render.gno:
```diff
@@ -0,0 +1,285 @@
+	}
+	out += md.BulletList(stats)
+
+	out += md.Blockquote("AwesomeGno is a community-driven project that aims to showcase and promote the best items built on gno.land. Each item is carefully categorized and reviewed to ensure quality and usefulness.")
```
done in https://github.com/gnolang/gno/pull/3963/commits/80e773d5431834f3d59c5999a84302832671f079

matijamarjanovic committed:
correct update item

matijamarjanovic committed:
update tests with cross/crossing

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
- remove my address from admins
- add some new lines in render
- remove the check for empty address since there is IsValid() check
- make authors clickable

leohhhn committed:
Merge branch 'master' into feat/awesome-gno

leohhhn committed:
Merge branch 'master' into feat/awesome-gno

leohhhn committed:
fix lint

leohhhn commented:
Can you add a function to render out links to the last X items? We can then add this to the home page.

matijamarjanovic committed:
add fifo lib to render widget to get 5 latests items

matijamarjanovic commented:
> Can you add a function to render out links to the last X items? We can then add this to the home page.

I have done so in https://github.com/gnolang/gno/pull/3963/commits/f8fafc65fd022a17da02c704d78316dd6dd7ebba; the problem I have found with this implementation (you suggested using fifo package) is that the admins would have to predefine the size of the items to get manually and get the list of that size. This means you couldn't technically dinamically change the size of the return values since the items in the list would be cut off if X is smaller than the current size or would just return current number of items if X is larger than that since the rest of the spaces would be empty.

If it is significant to have a dinamically changing value i would stick to using avl tree and just iterate it X times since the items are sorted in the order they were added chronologically anyway. Let me know which implementation is more suiting and I can do it. I have also used the same formatting in the "widget" as I have in awesome gno `Render()`

leohhhn commented on examples/gno.land/r/demo/awesomegno/widget.gno:
```diff
@@ -0,0 +1,67 @@
+)
+
+func GetLatestItemsWidget() string {
+	crossing()
```
this should be readonly so no need to make it crossing

matijamarjanovic replied:
okay, done in https://github.com/gnolang/gno/pull/3963/commits/c66bf29089d7767a2db02a9ca3d10cca5b034e56

leohhhn commented on examples/gno.land/r/demo/awesomegno/widget.gno:
```diff
@@ -0,0 +1,67 @@
+}
+
+func GetLatestCategoriesWidget() string {
+	crossing()
```
Same here, readonly so no need for crossing

matijamarjanovic replied:
done in https://github.com/gnolang/gno/pull/3963/commits/c66bf29089d7767a2db02a9ca3d10cca5b034e56

matijamarjanovic committed:
remove crossing from widget functions

thehowl committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
update hall of realms registration

gfanton committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
- change gno.mod to gnomod.toml
- implement new cross syntax
- update tests

matijamarjanovic committed:
update widget.gmo

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic committed:
Merge branch 'feat/awesome-gno' of https://github.com/matijamarjanovic/gno into feat/awesome-gno

matijamarjanovic committed:
chore: update imports

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

matijamarjanovic commented:
Hey @leohhhn, I've updated the imports in awesome gno, can you please re-review and let me know if there are any changes to be made

matijamarjanovic committed:
chore: update matija gnomod.toml

matijamarjanovic committed:
Merge branch 'feat/awesome-gno' of https://github.com/matijamarjanovic/gno into feat/awesome-gno

matijamarjanovic committed:
chore : add some emojis to render

matijamarjanovic committed:
chore: separate errors in their own file errors.gno

matijamarjanovic committed:
chore: fmt

matijamarjanovic committed:
Merge branch 'master' into feat/awesome-gno

thehowl commented:
Hi @matijamarjanovic — thanks for putting this together! Closing as part of a stale-PR housekeeping sweep; the branch has been inactive since Sep 2025. If there's still appetite for an awesome-gno list, feel free to reopen or restart the discussion.
